# Which Campaigns are Actually Worth the Spend?

### 1. Load the dataset.
Then check the basics:
- dimensionality (rows, columns)
- column names, datatypes, and number of non-null values
- descriptive statistics (for numerical columns)
- are there any missing values?

In [10]:
import pandas as pd
url = "https://raw.githubusercontent.com/super-unknowledge/campaign-analysis-1/refs/heads/main/synthetic_social_media_campaign_data.csv"
df = pd.read_csv(url)

df.head()

,campaign_id,campaign_name,platform,post_type,post_date,post_time,content_theme,target_audience,boosted_post,spend_usd,...,conversions,conversion_value_usd,cpc_usd,ctr,engagement_rate,hashtag_count,emoji_count,word_count,sentiment_score,video_length_seconds
0,CMP1000,Summer Campaign 2025,TikTok,Story,1/8/2024,13:45,Promotion,"Gen Z (18-24, Urban)",False,8.46,...,25,503.43,0.05,0.0297,0.0883,9,2,18,0.12,37
1,CMP1000,Back to School Campaign 2024,X,Video,1/3/2024,19:30,Behind-the-Scenes,"Parents (30-45, Suburban)",True,7.07,...,5,249.50,0.14,0.0098,0.0320,2,0,12,-0.16,22
2,CMP1000,Spring Campaign 2025,LinkedIn,Story,1/20/2024,9:30,User Testimonial,"Gen Z (18-24, Urban)",True,40.60,...,56,2138.28,0.11,0.0300,0.1152,13,5,16,0.30,113
3,CMP1000,Summer Campaign 2024,Facebook,Image,1/14/2024,11:00,User Testimonial,"Gen X (35-50, Mixed)",True,5.23,...,19,1360.52,0.03,0.0553,0.1167,1,9,45,-0.39,0
4,CMP1000,Spring Campaign 2025,Instagram,Video,1/4/2024,11:00,Behind-the-Scenes,"Millennials (25-34, Suburban)",False,5.38,...,10,1312.49,0.07,0.0244,0.0694,13,2,40,-0.11,119


In [11]:
df.shape

(1610, 24)

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1610 entries, 0 to 1609
Data columns (total 24 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   campaign_id           1610 non-null   object 
 1   campaign_name         1610 non-null   object 
 2   platform              1610 non-null   object 
 3   post_type             1610 non-null   object 
 4   post_date             1610 non-null   object 
 5   post_time             1610 non-null   object 
 6   content_theme         1610 non-null   object 
 7   target_audience       1610 non-null   object 
 8   boosted_post          1610 non-null   bool   
 9   spend_usd             1610 non-null   float64
 10  reach                 1610 non-null   int64  
 11  impressions           1610 non-null   int64  
 12  engagements           1610 non-null   int64  
 13  clicks                1610 non-null   int64  
 14  conversions           1610 non-null   int64  
 15  conversion_value_usd 

In [13]:
df.describe()

,spend_usd,reach,impressions,engagements,clicks,conversions,conversion_value_usd,cpc_usd,ctr,engagement_rate,hashtag_count,emoji_count,word_count,sentiment_score,video_length_seconds
count,1610.000000,1610.000000,1610.000000,1610.000000,1610.000000,1610.000000,1610.000000,1610.000000,1610.000000,1610.000000,1610.000000,1610.000000,1610.000000,1610.000000,1610.000000
mean,37.921410,5677.191925,7577.076398,397.242857,198.421118,24.656522,1947.463789,0.242106,0.025947,0.069577,7.454037,5.114907,27.573292,0.183379,38.767702
std,49.690521,3542.976873,4596.120964,315.775973,167.967635,24.441697,2435.286218,0.327396,0.013041,0.028743,4.549370,3.174023,13.118617,0.464015,40.813124
min,0.010000,166.000000,259.000000,4.000000,1.000000,0.000000,0.000000,0.000000,0.003400,0.019100,0.000000,0.000000,5.000000,-1.000000,0.000000
25%,4.962500,3226.250000,4328.750000,171.000000,80.000000,9.000000,471.580000,0.040000,0.015600,0.044600,3.000000,2.000000,16.000000,-0.140000,0.000000
50%,17.355000,5247.000000,7106.000000,326.000000,156.000000,18.000000,1123.225000,0.140000,0.024000,0.068700,7.500000,5.000000,27.000000,0.190000,24.000000
75%,52.972500,7428.750000,9823.750000,542.000000,267.750000,33.000000,2494.772500,0.320000,0.034275,0.094100,11.000000,8.000000,39.000000,0.510000,75.000000
max,425.390000,28457.000000,33893.000000,2684.000000,1337.000000,220.000000,21624.050000,3.650000,0.069900,0.119900,15.000000,10.000000,50.000000,1.000000,120.000000


In [14]:
print(df.isnull().sum())

campaign_id             0
campaign_name           0
platform                0
post_type               0
post_date               0
post_time               0
content_theme           0
target_audience         0
boosted_post            0
spend_usd               0
reach                   0
impressions             0
engagements             0
clicks                  0
conversions             0
conversion_value_usd    0
cpc_usd                 0
ctr                     0
engagement_rate         0
hashtag_count           0
emoji_count             0
word_count              0
sentiment_score         0
video_length_seconds    0
dtype: int64


Let's peek at the unique values of some of the columns we'll use later:

In [ ]:
# Check unique values for platform column
# 1. Get the number of unique platforms
unique_count = df['platform'].nunique()

# 2. Get the list of all unique platforms
unique_values = df['platform'].unique()

print(unique_count, unique_values)

5 ['TikTok' 'X' 'LinkedIn' 'Facebook' 'Instagram']


In [ ]:
# Check unique values for post_type column
# 1. Get the number of unique post types
unique_count = df['post_type'].nunique()

# 2. Get the list of all unique post types
unique_values = df['post_type'].unique()

print(unique_count, unique_values)

5 ['Story' 'Video' 'Image' 'Carousel' 'Reel']


In [ ]:
# Check unique values for content_theme column
# 1. Get the number of unique content themes
unique_count = df['content_theme'].nunique()

# 2. Get the list of all unique content themes
unique_values = df['content_theme'].unique()

print(unique_count, unique_values)

5 ['Promotion' 'Behind-the-Scenes' 'User Testimonial' 'Product Launch'
 'Lifestyle']


Takeaways: 
- The dataset has 1,610 rows and 24 columns
- `post_date` and `post_time` are listed as "object". They'll need to be converted into datetime later. But since we're focusing on figuring out which campaigns are worth the spend, we'll leave them as is for now since they're we don't need that information.
- There are no missing values, and nothing to really clean since this is a generated dataset (Hooray!)

### 2. Perform the initial exploration.
We want to find out the following metrics:
- ROAS (Return on Ad Spend) = Conversion Value / Spend
- Cost per Conversion = Spend / Conversions
- Conversion Rate = Conversions / Clicks
- Efficiency Score = rank campaigns by ROAS within each platform

In Part 2 when we create the pipeline, we'll use postgres as our DB. SQL query versions of each line are left as comments where applicable.

In [19]:
# Only look at rows where we spend money and have conversions
# WHERE spend_usd > 0 AND conversions > 0
clean = df[(df['spend_usd'] > 0) & (df['conversions'] > 0)].copy()

# ROAS = Conversion Value / Spend
# Derived metric: Revenue per dollar spent
clean['roas'] = clean['conversion_value_usd'] / clean['spend_usd']

# Cost per Conversion = Spend / Conversions
# We know from the blog post that cpc_usd is cost per click, not conversion
# Derived metric: How much did we pay for each conversion?
clean['cost_per_conversion'] = clean['spend_usd'] / clean['conversions']

# Conversion Rate = Conversions / Clicks
# Derived metric: What percentage of the people who clicked get converted?
clean['conversion_rate'] = clean['conversions'] / clean['clicks']

# Efficiency Score
# Derived metric: Which campaigns have the best return per platform?
# Rank by ROAS within each Platform (Highest ROAS gets Rank 1)
clean['efficiency_rank'] = clean.groupby('platform')['roas'].rank(ascending=False, method='min')

In [20]:
# Assuming we sorted the final DataFrame for clear reading:
df_sorted = clean.sort_values(by=['platform', 'efficiency_rank'])
print(df_sorted[['platform', 'campaign_name', 'roas', 'efficiency_rank']].to_string(index=False))

 platform                campaign_name          roas  efficiency_rank
 Facebook        Holiday Campaign 2024  28878.090909              1.0
 Facebook        Holiday Campaign 2025  11071.428571              2.0
 Facebook           Fall Campaign 2024  10996.256410              3.0
 Facebook Back to School Campaign 2025  10393.140845              4.0
 Facebook Back to School Campaign 2024   6438.107143              5.0
 Facebook         Spring Campaign 2025   5870.884615              6.0
 Facebook        Holiday Campaign 2024   4370.374046              7.0
 Facebook         Summer Campaign 2024   4309.238095              8.0
 Facebook         Spring Campaign 2025   3253.820690              9.0
 Facebook         Summer Campaign 2024   2590.857143             10.0
 Facebook         Summer Campaign 2024   2218.392857             11.0
 Facebook         Spring Campaign 2024   2174.661224             12.0
 Facebook         Spring Campaign 2024   1897.000000             13.0
 Facebook        Hol

### 3. Perform A/B testing.
The dataset has multiple Post Types (Image, Video, Carousel, Story). 

So let's answer the question: "Do video posts actually convert better than image posts, or does it just look that way?"

We can do an A/B test by running a two-sample t-test on conversion rates between post types:

In [ ]:
from scipy import stats

video = clean[clean['post_type'] == 'Video']['conversion_rate']
image = clean[clean['post_type'] == 'Image']['conversion_rate']

t_stat, p_value = stats.ttest_ind(video, image)
print(video.count())
print(image.count())
print(f"p-value: {p_value:.4f}")
print("Statistically significant!" if p_value < 0.05 else "No significant difference — could be noise")

326
305
p-value: 0.3002
No significant difference — could be noise


We cannot confidently say video outperforms image on this platform. The difference could be random variation with this sample size.

Let's try answering: "Do Instagram posts convert better than posts on X?"

In [34]:
ig = clean[clean['platform'] == 'Instagram']['conversion_rate']
x = clean[clean['platform'] == 'X']['conversion_rate']

t_stat, p_value = stats.ttest_ind(ig, x)
print(ig.count())
print(x.count())
print(f"p-value: {p_value:.4f}")
print("Statistically significant!" if p_value < 0.05 else "No significant difference — could be noise")

321
351
p-value: 0.0246
Statistically significant!


So, we can say that our posts on Instagram convert than our posts on X. That makes sense intuitively.